### PR TITLE
generate_parenthesis

### PR DESCRIPTION
**Pull request by RenliangShi**
rename generate_parenthesis to generate_parenthesis_v1, add generate_parenthesis_v2 and test.

### CODE REVIEW

File: algorithms/backtrack/generate_parenthesis.py [hunk 1]
```diff
@@ -27,3 +27,17 @@ def add_pair(res, s, left, right):
...
     add_pair(res, "", n, 0)
     return res
+
+
+def generate_parenthesis_v2(n):
+    def add_pair(res, s, left, right):
+        if left == 0 and right == 0:
+            res.append(s)
+        if left > 0:
+            add_pair(res, s+"(", left-1, right)
```
[the remark this answers is not in the document]
Comment (RenliangShi) [reply algorithms/backtrack/generate_parenthesis.py]: Ok, finished!

[the remark this answers is not in the document]
Comment (danghai) [reply algorithms/backtrack/generate_parenthesis.py]: Thanks. But we need to wait #381 merge first?

[the remark this answers is not in the document]
Comment (RenliangShi) [reply algorithms/backtrack/generate_parenthesis.py]: ok, no problem.

**Comment by danghai** [comment]
@RenliangShi Thanks merge!

**Pull request merged**